File: camera/camera-core/src/test/java/androidx/camera/core/impl/ConfigTest.java
/*
 * Copyright 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.camera.core.impl;

import static androidx.camera.core.impl.Config.OptionPriority.ALWAYS_OVERRIDE;
import static androidx.camera.core.impl.Config.OptionPriority.OPTIONAL;
import static androidx.camera.core.impl.Config.OptionPriority.REQUIRED;
import static androidx.camera.core.impl.ImageOutputConfig.OPTION_RESOLUTION_SELECTOR;

import static com.google.common.truth.Truth.assertThat;

import android.os.Build;

import androidx.camera.core.resolutionselector.AspectRatioStrategy;
import androidx.camera.core.resolutionselector.ResolutionFilter;
import androidx.camera.core.resolutionselector.ResolutionSelector;
import androidx.camera.core.resolutionselector.ResolutionStrategy;
import androidx.camera.testing.impl.fakes.FakeUseCaseConfig;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.internal.DoNotInstrument;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;

@RunWith(RobolectricTestRunner.class)
@DoNotInstrument
@org.robolectric.annotation.Config(minSdk = Build.VERSION_CODES.LOLLIPOP)
public class ConfigTest {
    private static final String OPTION_ID = "OptionID";
    private static final String KEY = "Key";

    @Test
    public void configCreateWithIdAndClass() {
        Config.Option<Integer> option = Config.Option.create(OPTION_ID, Integer.class);
        assertThat(option.getId()).isEqualTo(OPTION_ID);
        assertThat(option.getValueClass()).isEqualTo(Integer.class);
        assertThat(option.getToken()).isNull();
    }

    @Test
    public void configCreateWithIdAndClassAndKey() {
        Config.Option<Integer> option = Config.Option.create(OPTION_ID, Integer.class, KEY);
        assertThat(option.getId()).isEqualTo(OPTION_ID);
        assertThat(option.getValueClass()).isEqualTo(Integer.class);
        assertThat(option.getToken()).isEqualTo(KEY);
    }

    @Test
    public void optionPriorityIsDeclaredCorrectly() {
        Set<Config.OptionPriority> set = new HashSet<>();
        set.add(ALWAYS_OVERRIDE);
        set.add(OPTIONAL);
        set.add(REQUIRED);

        assertThat(Collections.min(set)).isEqualTo(ALWAYS_OVERRIDE);

        set.clear();
        set.add(ALWAYS_OVERRIDE);
        set.add(REQUIRED);
        assertThat(Collections.min(set)).isEqualTo(ALWAYS_OVERRIDE);

        set.clear();
        set.add(OPTIONAL);
        set.add(REQUIRED);
        assertThat(Collections.min(set)).isEqualTo(REQUIRED);

        set.clear();
        set.add(OPTIONAL);
        set.add(ALWAYS_OVERRIDE);
        assertThat(Collections.min(set)).isEqualTo(ALWAYS_OVERRIDE);
    }

    @Test
    public void noConflict_whenTwoValueAreALWAYSOVERRIDE() {
        assertThat(Config.hasConflict(ALWAYS_OVERRIDE, ALWAYS_OVERRIDE)).isFalse();
    }

    @Test
    public void hasConflict_whenTwoValueAreREQUIRED() {
        assertThat(Config.hasConflict(REQUIRED, REQUIRED)).isTrue();
    }

    @Test
    public void noConflict_whenTwoValueAreOPTIONAL() {
        assertThat(Config.hasConflict(OPTIONAL, OPTIONAL)).isFalse();
    }

    @Test
    public void noConflict_whenTwoValueAreOPTIONAL_REQUIRED() {
        assertThat(Config.hasConflict(OPTIONAL, REQUIRED)).isFalse();
    }

    @Test
    public void noConflict_whenTwoValueAreOPTIONAL_ALWAYSOVERRIDE() {
        assertThat(Config.hasConflict(OPTIONAL, ALWAYS_OVERRIDE)).isFalse();
    }

    @Test
    public void noConflict_whenTwoValueAreREQUIRED_OPTIONAL() {
        assertThat(Config.hasConflict(REQUIRED, OPTIONAL)).isFalse();
    }

    @Test
    public void noConflict_whenTwoValueAreREQUIRED_ALWAYSOVERRIDE() {
        assertThat(Config.hasConflict(REQUIRED, ALWAYS_OVERRIDE)).isFalse();
    }

    @Test
    public void noConflict_whenTwoValueAreALWAYSREQUIRED_OPTIONAL() {
        assertThat(Config.hasConflict(ALWAYS_OVERRIDE, OPTIONAL)).isFalse();
    }

    @Test
    public void noConflict_whenTwoValueAreALWAYSOVERRIDE_REQUIRED() {
        assertThat(Config.hasConflict(ALWAYS_OVERRIDE, REQUIRED)).isFalse();
    }

    @Test
    public void overrideResolutionSelectorCorrectly() {
        MutableOptionsBundle mergedConfig = MutableOptionsBundle.create();
        MutableConfig baseConfig = new FakeUseCaseConfig.Builder().getMutableConfig();
        baseConfig.insertOption(OPTION_RESOLUTION_SELECTOR,
                new ResolutionSelector.Builder().setAspectRatioStrategy(
                                AspectRatioStrategy.RATIO_4_3_FALLBACK_AUTO_STRATEGY)
                        .setResolutionStrategy(
                                ResolutionStrategy.HIGHEST_AVAILABLE_STRATEGY).build());
        MutableConfig extendedConfig = new FakeUseCaseConfig.Builder().getMutableConfig();
        ResolutionFilter resolutionFilter = (supportedSizes, rotationDegrees) -> null;
        extendedConfig.insertOption(OPTION_RESOLUTION_SELECTOR,
                new ResolutionSelector.Builder().setAspectRatioStrategy(
                                AspectRatioStrategy.RATIO_16_9_FALLBACK_AUTO_STRATEGY)
                        .setResolutionFilter(resolutionFilter).build());
        Config.mergeOptionValue(mergedConfig, baseConfig, extendedConfig,
                OPTION_RESOLUTION_SELECTOR);

        ResolutionSelector mergedResolutionSelector =
                mergedConfig.retrieveOption(OPTION_RESOLUTION_SELECTOR);

        assertThat(mergedResolutionSelector.getAspectRatioStrategy()).isEqualTo(
                AspectRatioStrategy.RATIO_16_9_FALLBACK_AUTO_STRATEGY);
        assertThat(mergedResolutionSelector.getResolutionStrategy()).isEqualTo(
                ResolutionStrategy.HIGHEST_AVAILABLE_STRATEGY);
        assertThat(mergedResolutionSelector.getResolutionFilter()).isEqualTo(resolutionFilter);
    }

}

File: xr/runtime/runtime/src/test/kotlin/androidx/xr/runtime/ConfigTest.kt
/*
 * Copyright 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.xr.runtime

import android.graphics.Bitmap
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.google.common.truth.Truth.assertThat
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class ConfigTest {

    @Test
    fun equals_sameInstance_returnsTrue() {
        val config =
            Config.Builder()
                .setPlaneTracking(PlaneTrackingMode.HORIZONTAL_AND_VERTICAL)
                .setHandTracking(HandTrackingMode.BOTH)
                .build()

        assertThat(config).isEqualTo(config)
    }

    @Test
    fun equals_sameConfig_returnsTrue() {
        val config1 =
            Config.Builder()
                .setPlaneTracking(PlaneTrackingMode.HORIZONTAL_AND_VERTICAL)
                .setHandTracking(HandTrackingMode.BOTH)
                .build()
        val config2 =
            Config.Builder()
                .setPlaneTracking(PlaneTrackingMode.HORIZONTAL_AND_VERTICAL)
                .setHandTracking(HandTrackingMode.BOTH)
                .build()

        assertThat(config1).isEqualTo(config2)
    }

    @Test
    fun equals_differentPlaneTracking_returnsFalse() {
        val config1 =
            Config.Builder().setPlaneTracking(PlaneTrackingMode.HORIZONTAL_AND_VERTICAL).build()
        val config2 = Config.Builder().setPlaneTracking(PlaneTrackingMode.DISABLED).build()

        assertThat(config1).isNotEqualTo(config2)
    }

    @Test
    fun equals_differentHandTracking_returnsFalse() {
        val config1 = Config.Builder().setHandTracking(HandTrackingMode.BOTH).build()
        val config2 = Config.Builder().setHandTracking(HandTrackingMode.DISABLED).build()

        assertThat(config1).isNotEqualTo(config2)
    }

    @Test
    fun equals_differentFaceTracking_returnsFalse() {
        val config1 = Config.Builder().setFaceTracking(FaceTrackingMode.BLEND_SHAPES).build()
        val config2 = Config.Builder().setFaceTracking(FaceTrackingMode.DISABLED).build()

        assertThat(config1).isNotEqualTo(config2)
    }

    @Test
    fun equals_differentDepthEstimation_returnsFalse() {
        val config1 =
            Config.Builder().setDepthEstimation(DepthEstimationMode.SMOOTH_AND_RAW).build()
        val config2 = Config.Builder().setDepthEstimation(DepthEstimationMode.DISABLED).build()

        assertThat(config1).isNotEqualTo(config2)
    }

    @Test
    fun equals_differentAnchorPersistence_returnsFalse() {
        val config1 = Config.Builder().setAnchorPersistence(AnchorPersistenceMode.LOCAL).build()

        val config2 = Config.Builder().setAnchorPersistence(AnchorPersistenceMode.DISABLED).build()

        assertThat(config1).isNotEqualTo(config2)
    }

    @Test
    fun equals_differentImageTracking_returnsFalse() {
        val augmentedImageDatabase =
            AugmentedImageDatabase().apply {
                addAugmentedImageDatabaseEntry(
                    mode = AugmentedImageDatabaseEntryMode.DYNAMIC,
                    bitmap = Bitmap.createBitmap(1, 1, Bitmap.Config.ARGB_8888),
                )
            }
        val config1 = Config.Builder().setAugmentedImageDatabase(AugmentedImageDatabase()).build()
        val config2 = Config.Builder().setAugmentedImageDatabase(augmentedImageDatabase).build()

        assertThat(config1).isNotEqualTo(config2)
    }

    @Test
    fun equals_differentQrCodeTracking_returnsFalse() {
        val config1 = Config(qrCodeTracking = QrCodeTrackingMode.DISABLED)
        val config2 = Config(qrCodeTracking = QrCodeTrackingMode.DYNAMIC)

        assertThat(config1).isNotEqualTo(config2)
    }

    @Test
    fun hashCode_sameConfig_returnsSameHashCode() {
        val config1 =
            Config.Builder()
                .setPlaneTracking(PlaneTrackingMode.HORIZONTAL_AND_VERTICAL)
                .setHandTracking(HandTrackingMode.BOTH)
                .build()
        val config2 =
            Config.Builder()
                .setPlaneTracking(PlaneTrackingMode.HORIZONTAL_AND_VERTICAL)
                .setHandTracking(HandTrackingMode.BOTH)
                .build()

        assertThat(config1.hashCode()).isEqualTo(config2.hashCode())
    }

    @Test
    fun hashCode_differentConfig_returnsDifferentHashCode() {
        val config1 =
            Config.Builder()
                .setPlaneTracking(PlaneTrackingMode.HORIZONTAL_AND_VERTICAL)
                .setHandTracking(HandTrackingMode.BOTH)
                .build()
        val config2 =
            Config.Builder()
                .setPlaneTracking(PlaneTrackingMode.DISABLED)
                .setHandTracking(HandTrackingMode.BOTH)

        assertThat(config1.hashCode()).isNotEqualTo(config2.hashCode())
    }

    @Test
    fun builder_constructedWithConfig_createsSameConfig() {
        val config =
            Config(
                planeTracking = PlaneTrackingMode.HORIZONTAL_AND_VERTICAL,
                handTracking = HandTrackingMode.BOTH,
            )
        val copy = Config.Builder(config).build()

        assertThat(copy).isEqualTo(config)
        assertThat(copy).isNotSameInstanceAs(config)
    }
}
